packages/tools/src/utilities/stackPrefetch/stackContextPrefetch.ts

```diff
@@ -0,0 +1,386 @@
+  lowEnd = Math.round(lowEnd) || 0;
+  highEnd = Math.round(highEnd) || 0;
+
+  const arr = [];
+  let c = highEnd - lowEnd + 1;
+
+  if (c <= 0) {
+    return arr;
+  }
+
+  while (c--) {
+    arr[c] = highEnd--;
+  }
+
+  return arr;
+}
+
+function getStackData(element) {
+  const enabledElement = getEnabledElement(element);
+
+  if (!enabledElement) {
+    throw new Error(
+      'stackPrefetch: element must be a valid Cornerstone enabled element'
+    );
+  }
+
+  const { viewport } = enabledElement;
+
+  if (!(viewport instanceof StackViewport)) {
+    throw new Error(
+      'stackPrefetch: element must be a StackViewport, VolumeViewport stackPrefetch not yet implemented'
+    );
+  }
+
+  return {
+    currentImageIdIndex: viewport.getCurrentImageIdIndex(),
+    imageIds: viewport.getImageIds(),
+  };
+}
+
+const clearFromImageIds = (stack) => {
+  const { imageIds } = stack;
+  const imageIdSet = new Set<string>();
+  imageIds.forEach((imageId) => imageIdSet.add(imageId));
+  return (requestDetails) =>
+    requestDetails.type !== requestType ||
+    !imageIdSet.has(requestDetails.additionalDetails.imageId);
+};
+
+function prefetch(element) {
+  const stack = getStackData(element);
+  if (!stack || !stack.imageIds || stack.imageIds.length === 0) {
+    console.warn('CornerstoneTools.stackPrefetch: No images in stack.');
+    return;
+  }
+
+  // Get the stackPrefetch tool data
+  const stackPrefetchData = getToolState(element);
+
+  if (!stackPrefetchData) {
+    return;
+  }
+
+  const stackPrefetch = stackPrefetchData || {};
+
+  // If all the requests are complete, disable the stackPrefetch tool
+  if (
+    !stackPrefetch.indicesToRequest ||
+    !stackPrefetch.indicesToRequest.length
+  ) {
+    stackPrefetch.enabled = false;
+  }
+
+  // Make sure the tool is still enabled
+  if (stackPrefetch.enabled === false) {
+    return;
+  }
+
+  // Remove an imageIdIndex from the list of indices to request
+  // This fires when the individual image loading deferred is resolved
+  function removeFromList(imageIdIndex) {
+    const index = stackPrefetch.indicesToRequest.indexOf(imageIdIndex);
+
+    if (index > -1) {
+      // Don't remove last element if imageIdIndex not found
+      stackPrefetch.indicesToRequest.splice(index, 1);
+    }
+  }
+
+  // Remove all already cached images from the
+  // IndicesToRequest array.
+  const indicesToRequestCopy = stackPrefetch.indicesToRequest.slice();
+  const { currentImageIdIndex } = stack;
+
+  indicesToRequestCopy.forEach((imageIdIndex) => {
+    const imageId = stack.imageIds[imageIdIndex];
+
+    if (!imageId) {
+      return;
+    }
+
+    const distance = Math.abs(currentImageIdIndex - imageIdIndex);
+    // Use getter for nearby objects to ensure the recent access is updated
+    // and is cached for distant objects, to let them be decached if necessary
+    const imageCached =
```

Comment:
I'm not sure if I follow this, you are saying if distance is less than 6 check if the image is requested to be loaded (since that is basically the getImageLoadObject, it does not gaurantee the pixel load, since it is a promise), but if the distance is more than 6 check if it is pixel loaded. right? why? can you add this clarfication and a bit more to the comment above?

Reply:
can you update the doc based on what we discussed